cns/deviceplugin/plugin.go

```diff
@@ -0,0 +1,143 @@
+package deviceplugin
+
+import (
+	"context"
+	"net"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+type Plugin struct {
+	Logger              *zap.Logger
+	ResourceName        string
+	SocketWatcher       *SocketWatcher
+	Socket              string
+	deviceCountMutex    sync.Mutex
+	deviceCount         int
+	kubeletSocket       string
+	deviceCheckInterval time.Duration
+}
+
+func NewPlugin(l *zap.Logger, resourceName string, socketWathcer *SocketWatcher, socket string,
+	initialDeviceCount int, kubeletSocket string, deviceCheckInterval time.Duration,
+) *Plugin {
+	return &Plugin{
+		Logger:              l.With(zap.String("resourceName", resourceName)),
+		ResourceName:        resourceName,
+		SocketWatcher:       socketWathcer,
+		Socket:              socket,
+		deviceCount:         initialDeviceCount,
+		kubeletSocket:       kubeletSocket,
+		deviceCheckInterval: deviceCheckInterval,
+	}
+}
+
+// Run runs the plugin until the context is cancelled, restarting the server as needed
+func (p *Plugin) Run(ctx context.Context) {
+	defer p.mustCleanUp()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			p.Logger.Info("starting device plugin for resource", zap.String("resource", p.ResourceName))
+			if err := p.run(ctx); err != nil {
+				p.Logger.Error("device plugin for resource exited", zap.String("resource", p.ResourceName), zap.Error(err))
+			}
+		}
+	}
+}
+
+func (p *Plugin) run(ctx context.Context) error {
+	childCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	s := NewServer(p.Socket, p.Logger, p, p.deviceCheckInterval)
+	runErrChan := make(chan error, 2) //nolint:gomnd // disabled in favor of readability
+	go func(errChan chan error) {
+		if err := s.Run(childCtx); err != nil {
+			errChan <- err
+		}
+	}(runErrChan)
+
+	// wait till the server is ready before registering with kubelet
+	readyErrChan := make(chan error, 2) //nolint:gomnd // disabled in favor of readability
+	go func(errChan chan error) {
+		errChan <- s.Ready(childCtx)
+	}(readyErrChan)
+
+	select {
+	case err := <-runErrChan:
+		return errors.Wrap(err, "error starting grpc server")
+	case err := <-readyErrChan:
+		if err != nil {
+			return errors.Wrap(err, "error waiting on grpc server to be ready")
+		}
+	case <-ctx.Done():
+		return nil
+	}
```

Comment:
as long as there is no ordering requirement, could simplify with errgroup
```suggestion
	g := errgroup.Group{}
	g.Go(func() error {
		return s.Run(childCtx)
	})
	g.Go(func() error {
		return s.Ready(childCtx)
	})
	if err := g.Wait(); err != nil {
		return errors.Wrap(err, "error running device plugin server")
	}
```

Reply:
errgroups can also take a context, which might help with handling of `cancel`.

Reply:
Good suggestion. I assumed `s.Run` and `s.Ready` would return when the context passed to them closes, but it would probably be more robust to pass it to the errgroup also
```go
	g, cancel := errgroup.WithContext(ctx)
	defer cancel()
	g.Go(func() error {
		return s.Run(ctx)
	})
	g.Go(func() error {
		return s.Ready(ctx)
	})
	if err := g.Wait(); err != nil {
		return errors.Wrap(err, "error running device plugin server")
	}
```

Reply:
Hmm. This does simply the code.  A correction to code above. This is the signature of errgroup.WithContext(ctx)
`func errgroup.WithContext(ctx context.Context) (*errgroup.Group, context.Context)
` 
Also seems like, we don’t explicitly call cancel or use defer cancel() because errgroup handles context cancellation internally. The errgroup will cancel its context if any of the goroutines return an error, which will propagate the cancellation to all other goroutines and the context used in the goroutines

Updated code accordingly.

Reply:
@aggarwal0009 is correct, `Run` we want to run "forever" (AKA until the context is cancelled), but `Ready` should return a nil error as soon as the grpc server is up and ready to listen to connections. This is a requirement of the k8s device plugin, you can only advertise yourself as a plugin once you're ready to accept grpc connections, otherwise kubelet will reject your registration. It's not very obvious and can be made more clear probably, but the original code returns `nil` on the `readyErrChan` when it is ready, breaking out of the `select`

Reply:
hm, that wasn't clear at all in the code so that's still a problem

Reply:
Nothing that some documentation won't fix. I can add appropriate documentation to make this clear.

Reply:
Added some documentation. Hopefully that makes it clear.